README
python-sha3 - SHA-3 (NIST competition version) extension for CPython
====================================================================

# Warning!

**This package implements the original Keccak hash function as selected by NIST
in 2012. It is incompatible with the final SHA-3 standard.**

On April 7, 2014, SHA-3 draft standard was released with a change that
introduced additional padding of input data. The updated specification became
final as FIPS 202 on August 5, 2015.
This Python module does not include the modification and will produce different
hashes!

There are no plans to update this module to the latest revision of the
standard. It will remain compliant with the original Keccak proposal.
To calculate actual SHA-3 hashes, please consider migrating your code to a
FIPS 202 implementation (like https://github.com/bjornedstrom/python-sha3).

This extension module can be used with Python 2.x (from 2.4 upwards) and 3.x.

Four classes with different digest sizes are currently available:
 - sha3.sha224 - SHA-3 224 bits
 - sha3.sha256 - SHA-3 256 bits
 - sha3.sha384 - SHA-3 384 bits
 - sha3.sha512 - SHA-3 512 bits

The interface of these classes follows conventions from Python's builtin
hashlib module.

Internally python-sha3 uses the optimized version of the Keccak algorithm,
written and released into the public domain by its designers.
Both 32-bit and 64-bit architectures are supported (tested on x86 and x86_64
architectures).

Like hashlib, python-sha3 has optional support for concurrency.
To enable concurrent hashing, pass '--enable-threads' when building the module
with setup.py.
When built with multithreading support, python-sha3 will release the GIL
during hash calculations and use a per object, fine-grained lock instead.

The most current development version is available at:
https://github.com/ajakubek/python-sha3/

Bugs can be reported at:
https://github.com/ajakubek/python-sha3/issues/

This software is distributed under the MIT license.
Please see the LICENSE file included in the package for details.


CHANGES
2015-12-02
 - fixed segfault on 64-bit machines due to type mismatch of
   PyArg_ParseTuple argument (issue #1)


tests/test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import gc
import sys
import unittest

gc.set_debug(gc.DEBUG_UNCOLLECTABLE | gc.DEBUG_STATS)

if sys.hexversion >= 0x03000000:
    test_names = [
        'py3_tests.test_sha_224',
        'py3_tests.test_sha_256',
        'py3_tests.test_sha_384',
        'py3_tests.test_sha_512',
    ]
else:
    test_names = [
        'py2_tests.test_sha_224',
        'py2_tests.test_sha_256',
        'py2_tests.test_sha_384',
        'py2_tests.test_sha_512',
    ]

def suite():
    suite = unittest.TestSuite()
    loader = unittest.TestLoader()
    suite.addTest(loader.loadTestsFromNames(test_names))
    return suite


if __name__ == '__main__':
    unittest.TextTestRunner(verbosity=2).run(suite())


tests/py2_tests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import binascii
import unittest
from sha3 import sha224, sha256, sha384, sha512

class test_sha3(unittest.TestCase):

    def test_init(self):
        self.assertNotEqual(self.sha(), None)

    def test_empty_hash(self):
        self.assertEqual(self.h.hexdigest(), self.empty_hexdigest)

    def test_str(self):
        self.assertTrue(isinstance(str(self.h), str))

    def test_repr(self):
        self.assertTrue(isinstance(repr(self.h), str))

    def test_digest_size(self):
        self.assertEqual(self.h.digest_size, self.digest_size)
        self.assertEqual(len(self.h.digest()), self.h.digest_size)

    def test_hexdigest_size(self):
        self.assertEqual(len(self.h.hexdigest()), self.digest_size * 2)

    def test_block_size(self):
        self.assertEqual(self.h.block_size, self.block_size)

    def test_rate(self):
        self.assertEqual(self.h.rate, self.rate)

    def test_capacity(self):
        self.assertEqual(self.h.capacity, self.capacity)

    def test_invalid_message(self):
        invalid_types = [None, 0, 1.0, [], set()]
        for message in invalid_types:
            self.assertRaises(TypeError, self.h.update, message)

    def test_vectors(self):
        for message, hex_digest in self.vectors:
            h = self.sha()
            h.update(message)
            self.assertEqual(h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(h.digest()), hex_digest)
            # make sure that subsequent finish calls do not modify digest
            self.assertEqual(h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(h.digest()), hex_digest)

    def test_vectors_extended(self):
        for message, hex_digest in self.vectors_extended:
            self.h.update(message)
            self.assertEqual(self.h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(self.h.digest()), hex_digest)
            # make sure that subsequent finish calls do not modify digest
            self.assertEqual(self.h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(self.h.digest()), hex_digest)

    def test_vectors_unicode(self):
        for message, hex_digest in self.vectors:
            h = self.sha()
            h.update(unicode(message))
            self.assertEqual(h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(h.digest()), hex_digest)
            # make sure that subsequent finish calls do not modify digest
            self.assertEqual(h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(h.digest()), hex_digest)

    def test_vectors_unicode_extended(self):
        for message, hex_digest in self.vectors_extended:
            self.h.update(unicode(message))
            self.assertEqual(self.h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(self.h.digest()), hex_digest)
            # make sure that subsequent finish calls do not modify digest
            self.assertEqual(self.h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(self.h.digest()), hex_digest)

    def test_copy(self):
        copy = self.h.copy()
        h_orig_digest = self.h.hexdigest()
        copy_orig_digest = copy.hexdigest()
        self.assertEqual(h_orig_digest, copy_orig_digest)
        self.assertEqual(self.h.digest_size, copy.digest_size)
        self.assertEqual(self.h.block_size, copy.block_size)
        self.assertEqual(self.h.rate, copy.rate)
        self.assertEqual(self.h.capacity, copy.capacity)
        for message, hex_digest in self.vectors_extended:
            self.h.update(message)
            self.assertEqual(copy.hexdigest(), copy_orig_digest)
            self.assertEqual(binascii.hexlify(copy.digest()), copy_orig_digest)
            self.assertEqual(self.h.hexdigest(), hex_digest)
            self.assertEqual(binascii.hexlify(self.h.digest()), hex_digest)


class test_sha_224(test_sha3):

    def setUp(self):
        self.sha = sha224
        self.h = sha224()
        self.digest_size = 224 / 8
        self.block_size = 1152 / 8
        self.rate = 1152
        self.capacity = 448
        self.empty_hexdigest = 'f71837502ba8e10837bdd8d365adb85591895602fc552b48b7390abd'
        self.vectors = [
            ('', 'f71837502ba8e10837bdd8d365adb85591895602fc552b48b7390abd'),
            ('abcd', 'be2ec6c1cce305a0ba88300bfcad0ab0b9f480f964be34b2cd253199'),
            ('abcdefgh', '6b1f16d8ad4d5f51fe46d837480011df14fa2864a89dc887ee3d7134'),
            ('a'*1024*1024, '9318ddf4f64d3b24162ac76957c4ae7ea7e90d2ebb7c40d1ab19fb72'),
            ('\0'*1024*1024, 'c9f87efa6f26a8cdad2afd78d876e13a018938e21c3b0c05d110585b'),
        ]
        self.vectors_extended = [
            ('', 'f71837502ba8e10837bdd8d365adb85591895602fc552b48b7390abd'),
            ('abcd', 'be2ec6c1cce305a0ba88300bfcad0ab0b9f480f964be34b2cd253199'),
            ('efgh', '6b1f16d8ad4d5f51fe46d837480011df14fa2864a89dc887ee3d7134'),
            ('a'*1024*1024, '0dc0c7ae839ed2bc898ee69e84ab71f10b046d84d98f65c2be699a1a'),
            ('\0'*1024*1024, 'd62e3fde25c81a0ccff379075ba3385148d0da2b163c14ca194cf28a'),
        ]

class test_sha_256(test_sha3):

    def setUp(self):
        self.sha = sha256
        self.h = sha256()
        self.digest_size = 256 / 8
        self.block_size = 1088 / 8
        self.rate = 1088
        self.capacity = 512
        self.empty_hexdigest = 'c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470'
        self.vectors = [
            ('', 'c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470'),
            ('abcd', '48bed44d1bcd124a28c27f343a817e5f5243190d3c52bf347daf876de1dbbf77'),
            ('abcdefgh', '48624fa43c68d5c552855a4e2919e74645f683f5384f72b5b051b71ea41d4f2d'),
            ('a'*1024*1024, 'f5f3e54ad3d703f8e9edfd7ce79341b1d9286a692fa6c13ff13ee6ea94dbf97d'),
            ('\0'*1024*1024, '7b6ff0a03e9c5a8e77a2059bf28e26a7f0e8d3939a7cfe2193908ad8d683be90'),
        ]
        self.vectors_extended = [
            ('', 'c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470'),
            ('abcd', '48bed44d1bcd124a28c27f343a817e5f5243190d3c52bf347daf876de1dbbf77'),
            ('efgh', '48624fa43c68d5c552855a4e2919e74645f683f5384f72b5b051b71ea41d4f2d'),
            ('a'*1024*1024, '5b43799db735425a91016fb82240b775337a8c92e58512bf107f61b8342a4b95'),
            ('\0'*1024*1024, 'bf7c1b61e4568cc09140b0543a1be49989b0b9145d15e1a395d57939d5642580'),
        ]

class test_sha_384(test_sha3):

    def setUp(self):
        self.sha = sha384
        self.h = sha384()
        self.digest_size = 384 / 8
        self.block_size = 832 / 8
        self.rate = 832
        self.capacity = 768
        self.empty_hexdigest = '2c23146a63a29acf99e73b88f8c24eaa7dc60aa771780ccc006afbfa8fe2479b2dd2b21362337441ac12b515911957ff'
        self.vectors = [
            ('', '2c23146a63a29acf99e73b88f8c24eaa7dc60aa771780ccc006afbfa8fe2479b2dd2b21362337441ac12b515911957ff'),
            ('abcd', '565c4034428749960ad82523596b0bb422986c1941463ce4678e81391c6eb3e47be5d85b9394217dc7d25dda8328f392'),
            ('abcdefgh', '3f57fa1fe45b9dbfdd3c0e07fe5807c2c70ee6935bd35b2cf35750b52b15bdbbde372d8c4aee50013326fec4d86af805'),
            ('a'*1024*1024, 'a31e9cc5636b078739005e5ba799adc81c00121c9155e754e84a0efcd9b48f5144abf40ab5cfeea7f0045e9076dfe547'),
            ('\0'*1024*1024, '0f2e8d8c47013f356a5d9efc5a754e2c826aa3b411e549ab193cc30a49b1a3e984ff7a065306ce30834e1176331ec8b7'),

        ]
        self.vectors_extended = [
            ('', '2c23146a63a29acf99e73b88f8c24eaa7dc60aa771780ccc006afbfa8fe2479b2dd2b21362337441ac12b515911957ff'),
            ('abcd', '565c4034428749960ad82523596b0bb422986c1941463ce4678e81391c6eb3e47be5d85b9394217dc7d25dda8328f392'),
            ('efgh', '3f57fa1fe45b9dbfdd3c0e07fe5807c2c70ee6935bd35b2cf35750b52b15bdbbde372d8c4aee50013326fec4d86af805'),
            ('a'*1024*1024, '0eb8a277dcec02b31fa3c0559da769a2b8d91bb4837946e06ad4aa0b06f93d3cd424c0e17d4e419033aa9c0107ca63f7'),
            ('\0'*1024*1024, 'ab1dcde267d73b300f72c1085232b64791c50dac240da2102dd159195406e8b0363cfff4c68bf0241d384f0d8ef7d351'),
        ]

class test_sha_512(test_sha3):

    def setUp(self):
        self.sha = sha512
        self.h = sha512()
        self.digest_size = 512 / 8
        self.block_size = 576 / 8
        self.rate = 576
        self.capacity = 1024
        self.empty_hexdigest = '0eab42de4c3ceb9235fc91acffe746b29c29a8c366b7c60e4e67c466f36a4304c00fa9caf9d87976ba469bcbe06713b435f091ef2769fb160cdab33d3670680e'
        self.vectors = [
            ('', '0eab42de4c3ceb9235fc91acffe746b29c29a8c366b7c60e4e67c466f36a4304c00fa9caf9d87976ba469bcbe06713b435f091ef2769fb160cdab33d3670680e'),
            ('abcd', 'e4a7e8f5572f4853ef26a862f31687c249b1cd7922df2aac1f4348d8ceef944c74d1949e3465704a5f3f89fb53e0dcce3ea142c90af04c84cc7e548f144f8f0b'),
            ('abcdefgh', 'c96950698dd2e6e2051637687d676a64bf7170908d69004cab008fb4d5d25d780be1e0ca503f947f07859dd477249787705ef813b64abb6477a22aa1fb908d1d'),
            ('a'*1024*1024, 'b978f7ddb14b67d6ab89bc659be206cf3438cefe386bd9e025a7f7706759a6c25be415aebfbda582a446a45cbd5da03ea20197263907a7b2fe5002d1c2bcbee2'),
            ('\0'*1024*1024, '696adf53f32a68f4f5d92b44c2b46127b05dd2f4c590d09314949a68dc73384b7ecbd371dc3ff5896d2c49ab69d906116bd047fc29fa7f426843011819440396'),
        ]
        self.vectors_extended = [
            ('', '0eab42de4c3ceb9235fc91acffe746b29c29a8c366b7c60e4e67c466f36a4304c00fa9caf9d87976ba469bcbe06713b435f091ef2769fb160cdab33d3670680e'),
            ('abcd', 'e4a7e8f5572f4853ef26a862f31687c249b1cd7922df2aac1f4348d8ceef944c74d1949e3465704a5f3f89fb53e0dcce3ea142c90af04c84cc7e548f144f8f0b'),
            ('efgh', 'c96950698dd2e6e2051637687d676a64bf7170908d69004cab008fb4d5d25d780be1e0ca503f947f07859dd477249787705ef813b64abb6477a22aa1fb908d1d'),
            ('a'*1024*1024, 'c909d64f3a3b448f008f10fe7e3363e42dc3c798a8ae70f0c20e727abbd01923bd98beb6a9e62494b432ce567887800c500e808d7d1c1e91bd023600ac184a80'),
            ('\0'*1024*1024, 'ba9e08b95ca58a296e74593614ffc9601c4d1911027d010b3a49b211577d73af7f60e9c5490309a1ee64c7bbcbf37a967b3aef4b3a126d107f75041a36359585'),
        ]


def suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(test_sha_224))
    suite.addTest(unittest.makeSuite(test_sha_256))
    suite.addTest(unittest.makeSuite(test_sha_384))
    suite.addTest(unittest.makeSuite(test_sha_512))
    return suite


if __name__ == '__main__':
    unittest.TextTestRunner(verbosity=2).run(suite())


src/sha3_module.c
/* Copyright (c) 2012 Adam Jakubek
 * Released under the MIT license (see attached LICENSE file).
 */

#include "common.h"
#include "sha3_types.h"

static PyMethodDef sha3_methods[] =
{ 
    { NULL }    /* sentinel */
};

#ifndef PyMODINIT_FUNC  /* declarations for DLL import/export */
#define PyMODINIT_FUNC void
#endif

#if PY_MAJOR_VERSION >= 3

static struct PyModuleDef sha3_moduledef = {
    PyModuleDef_HEAD_INIT,
    "sha3",                     /* m_name */
    "SHA-3 hashing functions.", /* m_doc */
    -1,                         /* m_size */
    sha3_methods,               /* m_methods */
    NULL,                       /* m_reload */
    NULL,                       /* m_traverse */
    NULL,                       /* m_clear */
    NULL,                       /* m_free */
};

PyMODINIT_FUNC PyInit_sha3(void)
{
    PyObject* m;

    if (!sha3_init_types())
        return NULL;

    m = PyModule_Create(&sha3_moduledef);

    sha3_register_types(m);

    return m;
}

#else

PyMODINIT_FUNC initsha3(void)
{
    PyObject* m;

    if (!sha3_init_types())
        return;

    m = Py_InitModule3("sha3", sha3_methods,
                       "SHA-3 hashing functions.");

    sha3_register_types(m);
}

#endif


src/sha3_types.c
/* Copyright (c) 2012 Adam Jakubek
 * Released under the MIT license (see attached LICENSE file).
 */

#include "common.h"

#ifdef ENABLE_THREADS
#include "pythread.h"
#endif
#include <structmember.h>
#include "keccak/KeccakNISTInterface.h"

#define MAX_HASH_BYTES          (512 / 8)
#define MIN_CONCURRENT_SIZE     (4096)

#ifndef Py_TYPE
#define Py_TYPE(o) ((o)->ob_type)
#endif


static PyTypeObject SHA224Type;
static PyTypeObject SHA256Type;
static PyTypeObject SHA384Type;
static PyTypeObject SHA512Type;

typedef struct
{
    PyObject_HEAD
    hashState hash_state;
    Py_ssize_t digest_size;
    Py_ssize_t block_size;
    Py_ssize_t rate;
    Py_ssize_t capacity;
#ifdef ENABLE_THREADS
    PyThread_type_lock lock;
#endif
} SHAObject;


static void sha_dealloc(SHAObject* self)
{
#ifdef ENABLE_THREADS
    if (self->lock != NULL)
        PyThread_free_lock(self->lock);
#endif
    PyObject_Del(self);
}

static PyObject* sha224_new(PyTypeObject* type,
                            PyObject* args,
                            PyObject* kwds)
{
    SHAObject* self;

    self = PyObject_New(SHAObject, type);
    if (self == NULL)
        return NULL;

    self->digest_size = 224 / 8;
    self->block_size = 1152 / 8;
    self->rate = 1152;
    self->capacity = 448;
#ifdef ENABLE_THREADS
    self->lock = NULL;
#endif

    if (Init(&self->hash_state, 224) != SUCCESS)
        return NULL;

    return (PyObject*)self;
}

static PyObject* sha256_new(PyTypeObject* type,
                            PyObject* args,
                            PyObject* kwds)
{
    SHAObject* self;

    self = PyObject_New(SHAObject, type);
    if (self == NULL)
        return NULL;

    self->digest_size = 256 / 8;
    self->block_size = 1088 / 8;
    self->rate = 1088;
    self->capacity = 512;
#ifdef ENABLE_THREADS
    self->lock = NULL;
#endif

    if (Init(&self->hash_state, 256) != SUCCESS)
        return NULL;

    return (PyObject*)self;
}

static PyObject* sha384_new(PyTypeObject* type,
                            PyObject* args,
                            PyObject* kwds)
{
    SHAObject* self;

    self = PyObject_New(SHAObject, type);
    if (self == NULL)
        return NULL;

    self->digest_size = 384 / 8;
    self->block_size = 832 / 8;
    self->rate = 832;
    self->capacity = 768;
#ifdef ENABLE_THREADS
    self->lock = NULL;
#endif

    if (Init(&self->hash_state, 384) != SUCCESS)
        return NULL;

    return (PyObject*)self;
}

static PyObject* sha512_new(PyTypeObject* type,
                            PyObject* args,
                            PyObject* kwds)
{
    SHAObject* self;

    self = PyObject_New(SHAObject, type);
    if (self == NULL)
        return NULL;

    self->digest_size = 512 / 8;
    self->block_size = 576 / 8;
    self->rate = 576;
    self->capacity = 1024;
#ifdef ENABLE_THREADS
    self->lock = NULL;
#endif

    if (Init(&self->hash_state, 512) != SUCCESS)
        return NULL;

    return (PyObject*)self;
}

static int sha_init(SHAObject* self, PyObject* args, PyObject* kwds)
{
    if (Init(&self->hash_state, (int)self->digest_size * 8) != SUCCESS)
    {
        PyErr_SetString(PyExc_RuntimeError, "failed to initialize hash state");
        return -1;
    }

    return 0;
}

static PyObject* sha_update(SHAObject* self, PyObject* other)
{
    unsigned char* buffer;
    Py_ssize_t size;
    HashReturn result;

    if (!PyArg_ParseTuple(other, "s#:update", &buffer, &size))
        return NULL;

#if ENABLE_THREADS
    if (self->lock == NULL && size >= MIN_CONCURRENT_SIZE)
        self->lock = PyThread_allocate_lock();

    if (self->lock != NULL && size >= MIN_CONCURRENT_SIZE)
    {
        Py_BEGIN_ALLOW_THREADS
        PyThread_acquire_lock(self->lock, 1);

        result = Update(&self->hash_state, buffer, (DataLength)size * 8);

        PyThread_release_lock(self->lock);
        Py_END_ALLOW_THREADS
    }
    else
        result = Update(&self->hash_state, buffer, (DataLength)size * 8);
#else
    result = Update(&self->hash_state, buffer, (DataLength)size * 8);
#endif

    if (result != SUCCESS)
    {
        PyErr_SetString(PyExc_RuntimeError, "failed to update hash state");
        return NULL;
    }

    Py_RETURN_NONE;
}

static PyObject* sha_digest(SHAObject* self)
{
    char digest[MAX_HASH_BYTES];
    hashState tmp;
    HashReturn result;

    assert(self->digest_size <= MAX_HASH_BYTES);

#if ENABLE_THREADS
    /* Use lock if it's already created, but don't create a new one.
     * This is usually the last method called.
     */
    if (self->lock != NULL)
    {
        Py_BEGIN_ALLOW_THREADS
        PyThread_acquire_lock(self->lock, 1);

        memcpy(&tmp, &self->hash_state, sizeof(hashState));
        result = Final(&tmp, (BitSequence*)digest);

        PyThread_release_lock(self->lock);
        Py_END_ALLOW_THREADS
    }
    else
    {
        memcpy(&tmp, &self->hash_state, sizeof(hashState));
        result = Final(&tmp, (BitSequence*)digest);
    }
#else
    memcpy(&tmp, &self->hash_state, sizeof(hashState));
    result = Final(&tmp, (BitSequence*)digest);
#endif

    if (result!= SUCCESS)
    {
        PyErr_SetString(PyExc_RuntimeError,
            "failed to finalize hash calculation");
        return NULL;
    }

#if PY_MAJOR_VERSION >=3
    return PyBytes_FromStringAndSize(digest, self->digest_size);
#else
    return PyString_FromStringAndSize(digest, self->digest_size);
#endif
}

static PyObject* sha_hexdigest(SHAObject* self)
{
    static const char* DIGITS = "0123456789abcdef";

    unsigned char digest[MAX_HASH_BYTES];
    char hex_digest[MAX_HASH_BYTES * 2];
    char *p;
    int i;
    hashState tmp;
    HashReturn result;

    assert(self->digest_size <= MAX_HASH_BYTES);

#if ENABLE_THREADS
    /* Use lock if it's already created, but don't create a new one.
     * This is usually the last method called.
     */
    if (self->lock != NULL)
    {
        Py_BEGIN_ALLOW_THREADS
        PyThread_acquire_lock(self->lock, 1);

        memcpy(&tmp, &self->hash_state, sizeof(hashState));
        result = Final(&tmp, (BitSequence*)digest);

        PyThread_release_lock(self->lock);
        Py_END_ALLOW_THREADS
    }
    else
    {
        memcpy(&tmp, &self->hash_state, sizeof(hashState));
        result = Final(&tmp, (BitSequence*)digest);
    }
#else
    memcpy(&tmp, &self->hash_state, sizeof(hashState));
    result = Final(&tmp, (BitSequence*)digest);
#endif

    if (result != SUCCESS)
    {
        PyErr_SetString(PyExc_RuntimeError,
            "failed to finalize hash calculation");
        return NULL;
    }
 
    p = hex_digest;
    for (i = 0; i < self->digest_size; ++i)
    {
        *p++ = DIGITS[(digest[i] / 16) & 0xf];
        *p++ = DIGITS[(digest[i] % 16) & 0xf];
    }

#if PY_MAJOR_VERSION >=3
    return PyUnicode_FromStringAndSize(hex_digest, self->digest_size * 2);
#else
    return PyString_FromStringAndSize(hex_digest, self->digest_size * 2);
#endif
}

static PyObject* sha_copy(PyObject* self)
{
    SHAObject* sha_self = (SHAObject*)self;
    SHAObject* new_sha;

    if (Py_TYPE(self) == &SHA224Type)
        new_sha = (SHAObject*)PyObject_New(SHAObject, &SHA224Type);
    else if (Py_TYPE(self) == &SHA256Type)
        new_sha = (SHAObject*)PyObject_New(SHAObject, &SHA256Type);
    else if (Py_TYPE(self) == &SHA384Type)
        new_sha = (SHAObject*)PyObject_New(SHAObject, &SHA384Type);
    else if (Py_TYPE(self) == &SHA512Type)
        new_sha = (SHAObject*)PyObject_New(SHAObject, &SHA512Type);

    if (new_sha == NULL)
        return NULL;

    memcpy(&new_sha->hash_state, &sha_self->hash_state, sizeof(hashState));
    new_sha->digest_size = sha_self->digest_size;
    new_sha->block_size = sha_self->block_size;
    new_sha->rate = sha_self->rate;
    new_sha->capacity = sha_self->capacity;
#ifdef ENABLE_THREADS
    new_sha->lock = NULL;
#endif

    return (PyObject*)new_sha;
}

static PyObject* sha_repr(PyObject* self)
{
    PyObject* str;

#if PY_MAJOR_VERSION >=3
    str = PyUnicode_FromFormat("<%s HASH object @ %p>",
        Py_TYPE(self)->tp_name, self);
#else
    str = PyString_FromFormat("<%s HASH object @ %p>",
        Py_TYPE(self)->tp_name, self);
#endif
    if (str == NULL)
    {
        PyErr_SetString(PyExc_RuntimeError, "failed to format string");
        return NULL;
    }

    return str;
}

static PyMethodDef SHAMethods[] =
{
    { "update", (PyCFunction)sha_update, METH_VARARGS,
      "Update the hash object with the string in argument." },
    { "digest", (PyCFunction)sha_digest, METH_NOARGS,
      "Return the digest of the strings passed to the update() method." },
    { "hexdigest", (PyCFunction)sha_hexdigest, METH_NOARGS,
      "Return the hexadecimal digest of the strings passed to the update() method." },
    { "copy", (PyCFunction)sha_copy, METH_NOARGS,
      "Return a copy of the hash object." },
    { NULL },   /* sentinel */
};

static PyMemberDef SHAMembers[] =
{
    { "digest_size", T_INT, offsetof(SHAObject, digest_size), READONLY,
      "The size of the resulting hash in bytes." },
    { "block_size", T_INT, offsetof(SHAObject, block_size), READONLY,
      "The internal block size of the hash algorithm in bytes." },
    { "rate", T_INT, offsetof(SHAObject, rate), READONLY,
      "The rate of sponge function in bits." },
    { "capacity", T_INT, offsetof(SHAObject, capacity), READONLY,
      "The capacity of sponge function in bits." },
    { NULL },   /* sentinel */
};

#ifndef PyVarObject_HEAD_INIT
    #define PyVarObject_HEAD_INIT(type, size) \
        PyObject_HEAD_INIT(type) size,
#endif

#define DefineSHAType(bits)                                     \
    static PyTypeObject SHA##bits##Type =                       \
    {                                                           \
        PyVarObject_HEAD_INIT(NULL, 0)                          \
        "sha3.sha"#bits,                /* tp_name */           \
        sizeof(SHAObject),              /* tp_basicsize */      \
        0,                              /* tp_itemsize */       \
        (destructor)sha_dealloc,        /* tp_dealloc */        \
        0,                              /* tp_print */          \
        0,                              /* tp_getattr */        \
        0,                              /* tp_setattr */        \
        0,                              /* tp_compare */        \
        sha_repr,                       /* tp_repr */           \
        0,                              /* tp_as_number */      \
        0,                              /* tp_as_sequence */    \
        0,                              /* tp_as_mapping */     \
        0,                              /* tp_hash */           \
        0,                              /* tp_call */           \
        sha_repr,                       /* tp_str */            \
        0,                              /* tp_getattro */       \
        0,                              /* tp_setattro */       \
        0,                              /* tp_as_buffer */      \
        Py_TPFLAGS_DEFAULT,             /* tp_flags */          \
        "SHA-3 "#bits" bit version",    /* tp_doc */            \
        0,                              /* tp_traverse */       \
        0,                              /* tp_clear */          \
        0,                              /* tp_richcompare */    \
        0,                              /* tp_weaklistoffset */ \
        0,                              /* tp_iter */           \
        0,                              /* tp_iternext */       \
        SHAMethods,                     /* tp_methods */        \
        SHAMembers,                     /* tp_members */        \
        0,                              /* tp_getset */         \
        0,                              /* tp_base */           \
        0,                              /* tp_dict */           \
        0,                              /* tp_descr_get */      \
        0,                              /* tp_descr_set */      \
        0,                              /* tp_dictoffset */     \
        (initproc)sha_init,             /* tp_init */           \
        0,                              /* tp_alloc */          \
        sha##bits##_new,                /* tp_new */            \
    };

DefineSHAType(224)
DefineSHAType(256)
DefineSHAType(384)
DefineSHAType(512)

#undef DefineSHAtype


int sha3_init_types(void)
{
    return
        ((PyType_Ready(&SHA224Type) == 0) &&
         (PyType_Ready(&SHA256Type) == 0) &&
         (PyType_Ready(&SHA384Type) == 0) &&
         (PyType_Ready(&SHA512Type) == 0))
        ? 1 : 0;
}

void sha3_register_types(PyObject* module)
{
    Py_INCREF(&SHA224Type);
    Py_INCREF(&SHA256Type);
    Py_INCREF(&SHA384Type);
    Py_INCREF(&SHA512Type);

    PyModule_AddObject(module, "sha224", (PyObject*)&SHA224Type);
    PyModule_AddObject(module, "sha256", (PyObject*)&SHA256Type);
    PyModule_AddObject(module, "sha384", (PyObject*)&SHA384Type);
    PyModule_AddObject(module, "sha512", (PyObject*)&SHA512Type);
}
